### PR TITLE
add `_is_fresh_start_epoch` attribute to fit_loop to avoid reload dl twice when training starts

### LINKED ISSUE #9502: Dataloader is reloaded twice after resuming from checkpoint
### Motivation

PyTorch Lightning reloads dataloader twice after resuming from checkpoint:
the first time reloads before train loop start [[link](https://github.com/PyTorchLightning/pytorch-lightning/blob/d2def369ec0ecdcf65acbe834c75cd2ff778d491/pytorch_lightning/trainer/trainer.py#L1122)];
The second reloads of default value for `reload_dataloaders_every_n_epochs` [[link](https://github.com/PyTorchLightning/pytorch-lightning/blob/6ff43cbff7a7b4ea75cd3d17b27aa7addf46421c/pytorch_lightning/loops/fit_loop.py#L202)]
This behavior could be problematic:
- This behavior could bring user confusion. They expect `reload_dataloaders_every_n_epochs = 1` forces the dataloaders to reload once every n epochs, but it loads multiple times sometimes. 
- The dataloader loading behavior is inconsistent between resuming from checkpoint and fresh training start. This could break the internal states of dataloader, which requires users to keep track of global states. 
- It may impact Fault tolerance training correctly captures the state of dataloader

### Pitch
Let's remove  `self.reset_train_val_dataloaders(model)` in `_run_train()` [[link](https://github.com/PyTorchLightning/pytorch-lightning/blob/d2def369ec0ecdcf65acbe834c75cd2ff778d491/pytorch_lightning/trainer/trainer.py#L1122)]
Because reset_*_dataloader() will be called in fit loop, there's no need to call it in trainer. 

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by ninginthecloud**
## What does this PR do?

Fixes #9502 

This PR aims to fix the bug mentioned in issue #9502. 
During fit loop, train_dataloader is loaded twice when resuming from checkpoint. This underlying behavior does not meet user's expectation. 
There are two options to fix this issue:
- **Option 1** (implemented in this PR): introduce an attribute `_is_fresh_start_epoch` in fit_loop to track if a new fit_loop is called. `_is_fresh_start_epoch` can be `True` when 
1. Trainer. fit(model), current_epoch = 0
2. Trainer.fit(model, dm), current_epoch != 0 but new dataloader/datamoudle is provided
3. trainer=Trainer(resume_from_checkpoint=ckpt), Trainer.fit(), current_epoch!=0 but Trainer.fit() resumes from checkpoint

- **Option 2**: remove `self.reset_train_val_dataloaders(model)` in `Trainer._run_train()` and set the reload logics in corresponding loops. When I worked on this option, I found a lot of other underlying assumptions:
1. `self.num_training_batches = 0` in `trainer._setup_on_init()`. This value should be initialized as `self.num_training_batches = float('inf')`, otherwise, if we remove dataloader reloading in `Trainer._run_train()` before calling fit_loop.run(), the `skip` flag from `fit_loop` will be true. which makes `fit_loop` falsely stops. 
2. we allow users to call `trainer.fit()` multiple times with different dataloaders. removing dataloader reloading in `Trainer._run_train()`, it marks `fit_loop` unable to continue fitting with new data. In this sense, we still need  attribute `_is_fresh_start_epoch` mentioned in Option 1. 
3. `reset_val_dataloader()` is called before fit_loop officially starts. Sometimes it's called if sanity_check is enabled. naively removing `self.reset_train_val_dataloaders(model)` in `Trainer._run_train()` breaks a lot of tests since our usage and assumption of loading `val_dataloder()` changed. As mentioned in issue #8738, it's better to provide more gradually control of reloading train and val dataloader seperately. 

In summary, this PR adopts the **option 1** to quickly fix this bug. In the next step, let's complete issue #8738 and then refactor dataloader reloading. 
 
## Before submitting

- [X] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [X] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1164,6 +1164,12 @@ def _run_train(self) -> None:
 
         self.reset_train_val_dataloaders(model)
 
+        # fresh start epoch can be under following situations:
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1167]: Hey @ninginthecloud.

Quick idea to investigate. What if we remove the `reset_train_val_dataloaders` from the Trainer #1165 and left only the loop handle the reloading logic. This is a blocker for loop customization.

Best,
T.C

Comment (ninginthecloud) [reply pytorch_lightning/trainer/trainer.py:1167]: thanks for the suggestion, @tchaton ! That's  what I've mentioned in the description as option 2 for resolving this fix. We definitely can go with that route,and it will change a lot of existing underlying assumptions and behaviors. 

> **Option 2**: remove `self.reset_train_val_dataloaders(model)` in `Trainer._run_train()` and set the reload logics in corresponding loops. When I worked on this option, I found a lot of other underlying assumptions we need to fix:
> 1. `self.num_training_batches = 0` in `trainer._setup_on_init()`. This value should be initialized as `self.num_training_batches = float('inf')`, otherwise, if we remove dataloader reloading in `Trainer._run_train()` before calling fit_loop.run(), the `skip` flag from `fit_loop` will be true. which makes `fit_loop` falsely stops. 
> 2. we allow users to call `trainer.fit()` multiple times with different dataloaders. removing dataloader reloading in `Trainer._run_train()`, it marks `fit_loop` unable to continue fitting with new data. In this sense, we still need  attribute `_is_fresh_start_epoch` mentioned in Option 1. 
> 3. `reset_val_dataloader()` is called before fit_loop officially starts. Sometimes it's called if sanity_check is enabled. naively removing `self.reset_train_val_dataloaders(model)` in `Trainer._run_train()` breaks a lot of tests since our usage and assumption of loading `val_dataloder()` changed. As mentioned in issue #8738, it's better to provide more gradually control of reloading train and val dataloader seperately. 
> In summary, this PR adopts the option 1 to quickly fix this bug. In the next step, let's complete issue #8738 and then refactor dataloader reloading.

We agree that ultimately we want to provide loop customization, after all, it does not make sense to let trainer control  data reloading. Let me know what you think. I can either 1) postpone issue #8738 to complete option 2, or adopt option 1 and complete #8738 and then refactor.

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1167]: IMO if #9502 isn't blocking something important, we can go the longer route for option 2 and wait this out.

Comment (ninginthecloud) [reply pytorch_lightning/trainer/trainer.py:1167]: great~ Thanks @carmocca and @tchaton~ Let me update this PR soon.

Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py:1167]: Thanks @ninginthecloud

**Comment by ninginthecloud** [comment]
I will close this PR. Please move all discussion to the latest PR #9671, which implemented option 2.
